### ISSUE TITLE
Feature request: Support asynchronous iterators as return values from resolvers

### ISSUE THREAD

**Issue by asishallab** [graphql/graphql-js#1537]
So far resolvers are expected to return an "`Iterable`" object (`Symbol.iterator`). But many I/O packages in Node.js work with asynchronous Promises, as does e.g. the popular [Sequelize relational database ORM](http://docs.sequelizejs.com/). Enabling asynchronous iterators would elegantly facilitate reading very large amounts of records, e.g. from a relational database, and return them without causing heavy load on memory. See [this Stackoverflow question](https://stackoverflow.com/questions/52577461/in-the-graphql-reference-implementation-resolvers-are-expected-to-return-an-iter) for a more detailed discussion of the example. So far in such a case the only solution is standard pagination, which forces the client to implement batch wise loading of large data-sets. Especially when client and provider are maintained by different organisations, this can lead to issues.  
Could support for asynchronous iterators (`Symbol.asyncIterator`) be implemented in graphql-js, please?

**Comment by asishallab** [comment]
@Cito thank you for your quick reply and insights. Indeed I am talking about resolvers that return one or more objects defined as types in the respective GraphQL schema. These resolvers could of course be both root resolvers or those resolving List fields. Your suggested solution to return arrays of Promises indeed works, but might be technically inefficient or even constate an anti-pattern. For a good example please see the [provided Question / Example](https://stackoverflow.com/questions/52577461/in-the-graphql-reference-implementation-resolvers-are-expected-to-return-an-iter) on Stackoverflow that I mentioned above. It explains the predicament quite in detail, I hope.

As to your concern that async iterators not guarantee to finish, couldn't this be solved by a adequate time out? 

I am convinced that this feature request is pretty important in a wide variety of applications (among others, possibly Issue #1535), as so many I/O packages in node do not provide streams but work with Promises. The [Sequelize ORM](http://docs.sequelizejs.com/) is just one very popular example (see the above Stackoverflow question for details). While object-streams _can_ be converted into _synchronous_ iterators they are lamentably often _not_ supported by the respective I/O packages, as e.g. is the case with Sequelize. Once streams are not available and single record fetches constate an anti-pattern, due to the nature of the underlying I/O devices, efficient batch-wise fetching becomes a necessity. However, when wanting to avoid that the client is obligated to paginate through all results when requesting all available records, the batch wise fetching needs to be wrapped into an object-stream or, where no streams are available, an asynchronous iterator to still facilitate a so to speak "_read-all GraphQL query_" which does not cause high memory load or even out-of-memory crashes.

**Comment by Cito** [comment]
@asishalla It's still not clear to me whether you're speaking about ordinary requests or subscriptions. Regarding the former, you can only send multiple values for list fields, and I believe it's the responsibility of the resolver function to collect batches as arrays. Then you have full control how these batches are generated. Regarding subscriptions, async iterators are already supported as return values of subscribe functions. My issue in #1535 is only that the resolver function (which is executed for every single value) cannot be async.

**Comment by asishallab** [comment]
@Cito thank you again for your patience and support. 

First of all let me clarify, that I am only talking about ordinary requests and _not_ subscriptions. I apologize for the confusion.

Second I'd like to disclaim that I am still pretty new to GraphQL and hope not to have misunderstood the terminology. Probably a simplified example helps explaining what I would like to do.

In human language I have a relational database with two tables (data models) Dogs and Fleas. Obviously a dog might have many fleas, and a flea might only have a single dog. I access this database using the popular [Sequelize](http://docs.sequelizejs.com/) framework, and would like to have a GraphQL API for the two data models.
Now, because a dog might have millions of fleas, and we have millions of dogs, I'd like to serve large result sets in a batched manner without causing high memory load on the server and above all _without_ forcing the client to use pagination. So, my idea, after quite some experimentation, was to use a custom Iterator.

TL;DR: 
I apologize for the volume of code that follows.
;-)

```
const BATCH_MAX_RECORDS = 10000

findAllIterator: async function(model, options) {
    let nRequested = options.limit || await model.count(options)
    options.offset = options.offset || 0
    // The following limit Works also, if count < BATCH_MAX_RECORDS and if
    // original limit !== BATCH_MAX_RECORDS
    options.limit = BATCH_MAX_RECORDS
    // Return an instance of iterator
    return {
      [Symbol.asyncIterator]: function() {
        return this
      },
      model: model,
      nRequested: nRequested,
      options: options,
      batch: [],
      next: async function() {
        if (this.batch.length === 0 &&
          (this.options.offset + this.options.limit < this.nRequested +
            BATCH_MAX_RECORDS)) {
          this.batch = await this.model.findAll(options)
          this.options.offset = this.options.offset +
            BATCH_MAX_RECORDS
          console.log(
            `Loaded ${this.batch.length} records. Current offset incremented to ${this.options.offset}`
          )
        }
        return (this.batch.length === 0) ? {
          done: true,
          value: null
        } : {
          done: false,
          value: this.batch.shift()
        }
      }
    }
  }
``` 

The above works fine and fulfills both my objectives: It accesses the database batch-wise and returns an iterator to all records without causing high memory load.
I'd like to use it in the following two resolvers. If, I'm not mistaken the dog-resolver is a root resolver, and the flea-resolver is a List field resolver?

Schema gist
```  
type Dog {
  name: String
  fleas: [Flea]
}

type Flea {
  species: String
  dog: Dog
}
```

Resolver gist
```
// root resolver
allDogs(context) {
   // This is where I create the memory saving 
   // iterator that loads millions of dogs one batch at a time
   return findAllIterator( sequelize.models.Dog, {} )
}

// List field resolver
Dog.prototype.fleas(context) {
   // Memory friendly access to millions of fleas
   return findAllIterator( sequelize.models.Flea, {where: { dog_id: this.dog_id } } )
}
```

I hope this clears the fog?

**Comment by Cito** [comment]
@asishallab Thanks for the clarification. The crucial point is: Do you want to serve the list of all dogs or fleas of a dog in one go to the client, or also in multiple batches? If you want to serve everything at once, JavaScript will need to build and pass the large result JSON object anyway, even when you query the database with an iterator, which can cause the memory problems you want to avoid. So you need to query and serve in multiple batches, i.e. you can't avoid pagination. In order to do that, you would need to add GraphQL arguments to your fleas and allDogs fields so you can specify the offset and limit of the batches, like [here](https://graphql.org/learn/pagination/). The resolver would then pass your arguments to your iterator (actually a normal function would do) and send only that one batch as a List ("fleas" is specified as a List, and the "allDogs" root query field would be specified as a List as well). This is of course a very common problem, and [Relay Cursor Connections](https://facebook.github.io/relay/graphql/connections.htm) have been proposed as a standardized solution. The "cursors" keep track of the position in your iterator on the server side.

**Comment by asishallab** [comment]
@Cito Thank you very much for the clarification. It has been an eye opener, indeed.

So then all my efforts of reducing memory load on the server with any type of iterator are in vain, anyway.

I guess, what I'd really like to do is using an object-stream which - by GraphQL - is piped into a transform stream converting each object to JSON. However, I am unsure whether List fields would still render this approach futile?

Consider this example in pseudo-code
```
// root resolver
allDogs(context) {
  return dbCon('dogs').select('*').stream()
}

// List field resolver
Dog.prototype.fleas(context) {
   return dbCon('fleas').where('dog_id': this.id).stream()
}
```

Then GraphQL would convert each object in the stream to JSON and write that into the HttpResponse. For nested List Fields this would have to be a bit messy. The part of the dog until the fleas array would have to be written into the response stream and then each flea from the resolved List Field object-stream in turn. I hope I am able to explain myself?

chunk 1: { "name": "cool dog", fleas: [
chunk 2: { "species": "nasty flea" },
chunk 3: { "species": "another flea"},
...
last chunk: ]}

Wouldn't that work?

**Comment by Cito** [comment]
I think GraphQL with its nested and parallel resolve functions is simply not intended to be used as a streaming API, at least normal GraphQL queries are not meant to be used like that. But you could achieve something like this using a subscription. In your case you would subscribe to a stream of fleas, and then get your individual fleas or batches of fleas delivered one by one.

**Comment by asishallab** [comment]
@Cito Thank you for your help on all this. 

I guess we can close this Issue now. Just one more thing. Do you have a good documentation recommendation for GraphQL subscriptions? So far I can only find [this blog](https://graphql.org/blog/subscriptions-in-graphql-and-relay/). As mentioned before I'd like avoiding the usage of third party frameworks like Apollo and remain within the reference implementation.

**Comment by IvanGoncharov** [comment]
@Cito Thanks for handling this issue 👍 

> I guess we can close this Issue now.

@asishallab Can I close this issue?

**Comment by asishallab** [comment]
Thanks for this fruitful discussion. I learned a lot.
